zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/callback/ZkAsyncCallbacks.java

```diff
@@ -120,6 +122,39 @@ public void handle() {
     }
   }
 
+  public static class SyncCallbackHandler extends DefaultCallback implements AsyncCallback.VoidCallback {
+    private String _sessionId;
+
+    public SyncCallbackHandler(String sessionId) {
+      _sessionId = sessionId;
+    }
+
+    @Override
+    public void processResult(int rc, String path, Object ctx) {
+      LOG.info("sycnOnNewSession with sessionID {} async return code: {}", _sessionId, rc);
+      callback(rc, path, ctx);
+    }
+
+    @Override
+    public void handle() {
+      // Make compiler happy, not used.
+    }
+
+    @Override
+    protected boolean needRetry(int rc) {
+      try {
+        // Connection to the server has been lost
+        if (KeeperException.Code.get(rc) == Code.CONNECTIONLOSS) {
+          return true;
+        } 
+        return false;
+      } catch (ClassCastException | NullPointerException ex) {
```

Comment:
I wonder how `NullPointerException` is thrown?

Reply:
Copied from base class. @jiajunwang, what is your take?

Reply:
Haven't you check the reason before copy the code, Kai : )

This method may return null.

Reply:
KeeperException.Code.get() can throw null point exception.

Reply:
@kaisun2000 Could you help understand this: `KeeperException.Code.get() can throw null point exception.` Underneath the lookup is a map. map.get(rc) returns null if rc doesn't exist.
```
/**
         * Get the Code value for a particular integer error code
         * @param code int error code
         * @return Code value corresponding to specified int code, or null
         */
        public static Code get(int code) {
            return lookup.get(code);
        }
```

The code should be like
```
if (KeeperException.Code.get(rc) == Code.CONNECTIONLOSS) {
     return true;
}
LOG.error("Session {} failed to handle unknown return code {}. Skip retrying.",
             _sessionId, rc);
return false;
```

Reply:
@kaisun2000 I just realized you said the code is copied from https://github.com/apache/helix/blob/9e6dafa1f4151613f5537f4a49e18254620bc40d/zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/callback/ZkAsyncCallbacks.java#L255-L266

Let me explain: `Code.get(rc)` could return `null`, and `switch(null)` gives NPE. That's why the code needs to catch NPE. But for your code here, I don't think NPE would happen.

Reply:
This is a good point. I revert back to switch() same as JJ's original style. Logging unknown code due to NPE is important for us to troubleshooting if it happens in the field.